mindmaps-migration/json/src/main/java/io/mindmaps/migration/json/JsonDataMigrator.java

```diff
@@ -0,0 +1,155 @@
+/*
+ * MindmapsDB - A Distributed Semantic Database
+ * Copyright (C) 2016  Mindmaps Research Ltd
+ *
+ * MindmapsDB is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * MindmapsDB is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with MindmapsDB. If not, see <http://www.gnu.org/licenses/gpl.txt>.
+ */
+
+package io.mindmaps.migration.json;
+
+import io.mindmaps.MindmapsGraph;
+import io.mindmaps.concept.*;
+import io.mindmaps.concept.ResourceType.DataType;
+import mjson.Json;
+
+import java.util.UUID;
+
+/**
+ * Migrator for migrating JSON data into Mindmaps instances
+ */
+public class JsonDataMigrator {
+
+    private MindmapsGraph graph;
+
+    /**
+     * Create a JsonDataMigrator to migrate into the given graph
+     */
+    public JsonDataMigrator() {}
+
+    public JsonDataMigrator graph(MindmapsGraph graph){
+        this.graph = graph;
+        return this;
+    }
+
+
+    /**
+     * Migrate JSON data into Mindmaps instances
+     * @param name the name of the root concept type
+     * @param data the data to load into Mindmaps
+     * @return the instance representing the data
+     */
+    public Instance migrateData(String name, Json data) {
+        Type type = graph.getType(name);
+
+        return migrateData(type, data);
+    }
+
+    private Instance migrateData(Type type, Json data) {
+        JsonType jsonType = JsonType.getType(data);
+
+        // If the type has subtypes, then we choose the correct one based on the JSON data type
+        if (type.subTypes().size() > 1) {
+            type = graph.getType(JsonMapper.subtypeName(type, jsonType));
+        }
+
+        boolean isResource = type.isResourceType();//|| type.subTypes().stream().anyMatch(Type::isResourceType);
+        Instance instance = isResource ?
+                migrateResource(jsonType, type.asResourceType(), data) : migrateEntity(type.asEntityType());
+
+        switch (jsonType) {
+            case OBJECT:
+                migrateJsonObject(instance, data);
+                break;
+            case ARRAY:
+                migrateJsonArray(instance, data);
+                break;
+            case NULL:
```

Comment:
I don't think you need this case since its empty. It will just be ignored anyway.